### PR TITLE
[SYCL][PI] Move files to libpi folder

### PR DESCRIPTION
**Pull request by ProGTX**
The PI plugins could be made into an independent library,
allowing code outside the `sycl` project to use it.
The first step towards getting that
is to group all PI related files into its own directory structure.

This patch moves all the PI files into the folder `libpi`.
Includes setting up includes as `pi/piXXX`.
On the SYCL runtime side the files have been moved
from `pi.hpp` to `pi_sycl.hpp` etc.

Apart from the folder change and adjusting CMake,
there should be no change in functionality.
Tests had to be updated to set up include folders.

### CODE REVIEW

**Review by smaslov-intel — COMMENTED** [review]
@ProGTX : please update the description of PR with the motivation for this change.

**Comment by ProGTX** [comment]
> @ProGTX : please update the description of PR with the motivation for this change.

Done, the motivation is to eventually make it into an independent library so that other projects apart from `sycl` can use it.

**Comment by smaslov-intel** [comment]
> the motivation is to eventually make it into an independent library so that other projects apart from `sycl` can use it.

Could elaborate on what kind of uses you envision and why "libpi" is deemed to be the right abstraction layer for those uses?

File: sycl/include/CL/sycl/detail/array.hpp [hunk 1]
```diff
@@ -9,6 +9,8 @@
 #pragma once
 #include <CL/sycl/detail/type_traits.hpp>
 #include <CL/sycl/exception.hpp>
+#include <pi/pi.h>
```
Comment (bader) [inline sycl/include/CL/sycl/detail/array.hpp:12]: It's a pity that we had to include `pi.h` here to use `PI_INVALID_VALUE` in the function that has no dependency on plug-in interface.
```c++
  __SYCL_ALWAYS_INLINE void check_dimension(int dimension) const {
#ifndef __SYCL_DEVICE_ONLY__
    if (dimension >= dimensions || dimension < 0) {
      throw cl::sycl::invalid_parameter_error("Index out of range",
                                              PI_INVALID_VALUE);
    }
#endif
    (void)dimension;
  }
```
@romanovvlad, @smaslov-intel, I think there should be PI-independent error codes to report errors for cases like this.

File: sycl/include/CL/sycl/detail/pi_sycl.hpp [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// ===--------------------------------------------------------------------=== //
+
+#pragma once
+
+#include <CL/sycl/detail/defines.hpp>
+#include <pi/pi.hpp>
```
Comment (bader) [inline sycl/include/CL/sycl/detail/pi_sycl.hpp:12]: I'm not sure I understand the motivation to have this wrapper.
Do we have a dependency on `defines.hpp` from `pi.hpp`? If so, I think we should remove such tight coupling between plug-in interface and the rest of the sycl runtime library.

File: sycl/libpi/plugins/level_zero/CMakeLists.txt [hunk 3]
```diff
@@ -97,9 +97,10 @@ target_include_directories(LevelZeroLoader-Headers
 
 
 include_directories("${sycl_inc_dir}")
+include_directories("${pi_include_dir}")
```
Comment (bader) [inline sycl/libpi/plugins/level_zero/CMakeLists.txt:100]:
```suggestion
include_directories("${sycl_inc_dir}" "${pi_include_dir}")
```

File: sycl/source/CMakeLists.txt [hunk 4]
```diff
@@ -112,7 +113,7 @@ set(SYCL_SOURCES
     "detail/builtins_integer.cpp"
     "detail/builtins_math.cpp"
     "detail/builtins_relational.cpp"
-    "detail/pi.cpp"
+    "${pi_source_dir}/pi.cpp"
```
Comment (bader) [inline sycl/source/CMakeLists.txt:116]: Do I understand correctly that building libpi as a separate library will be added in follow-up patches?

File: sycl/test/basic_tests/vectors/ctad_fail.cpp [hunk 5]
```diff
@@ -1,4 +1,4 @@
-// RUN: %clangxx -std=c++17 -fsyntax-only -Xclang -verify %s -I %sycl_include/sycl -Xclang -verify-ignore-unexpected
+// RUN: %clangxx -fsyntax-only -Xclang -verify %s -fsycl -fsycl-targets=%sycl_triple -Xclang -verify-ignore-unexpected
```
Comment (bader) [inline sycl/test/basic_tests/vectors/ctad_fail.cpp:1]: How is this change related to moving files to libpi directory?
NOTE: I don't think we should make this change. Original test validates compilation of DPC++ header with C++ compiler, but this change involves device compilation as well.

File: sycl/unittests/kernel-and-program/Cache.cpp [hunk 6]
```diff
@@ -8,7 +8,7 @@
 
 #define SYCL2020_DISABLE_DEPRECATION_WARNINGS
 
-#include "CL/sycl/detail/pi.h"
+#include "CL/sycl/detail/pi_sycl.hpp"
```
Comment (bader) [inline sycl/unittests/kernel-and-program/Cache.cpp:11]:
```suggestion
#include "pi/pi.h"
```
?

File: sycl/include/CL/sycl/detail/array.hpp (hunk 1 above)
Comment (romanovvlad) [reply sycl/include/CL/sycl/detail/array.hpp:12]: Such error codes getting introduced in SYCL2020

**Comment by smaslov-intel** [comment]
> > the motivation is to eventually make it into an independent library so that other projects apart from `sycl` can use it.
> 
> Could elaborate on what kind of uses you envision and why "libpi" is deemed to be the right abstraction layer for those uses?

@ProGTX : friendly ping
